### PR TITLE
Speed up scoring procedure

### PR DESCRIPTION
**Pull request by nmdefries**
### Description
Speeds up scoring procedure by 5x. Modestly improves memory usage (-10%).

Test failure is related to an Ubuntu upgrade, addressing that separately.

### Changes
- Use plain R checks instead of `assert_that`. The assertions in `score_func_param_checker` make up a majority of the time spent producing scores when using `assertthat`; plain R is much faster.
- Use a direct `group_by`-`summarize` call to the list of error functions to avoid the `group_split` and nested `lapply`s. `erm` function is no longer needed.
- Avoid unnecessary `sort`s in `is_symmetric`. Checking if the quantiles are sorted is ~5x faster than sorting an already-sorted vector again.
- Avoid unnecessary `unique`s in the error measure definitions. Every error measure fn calls `score_func_param_checker` as a first step. `score_func_param_checker` checks that the group contains only one unique `actual`, so if that check passes, we can just take the first `actual` instead of calling `unique` again.

### CODE REVIEW

**Review by brookslogan — APPROVED** [review]
Think this is good to merge as-is, but have a few low-priority suggestions and questions.
* You mentioned considering moving from `assertthat` to the `checkmate` package; what made you go with plain R instead?
* What data/scripts were your benchmarks/profiling based on?  Is the benchmarking code easy to share here in case we want to re-use it for future performance experiments?
* The `length(actual_value) == length(value)` check is a bit weird.  I'm not sure if it's avoidable; the ~"evalcast-killcards" branch specifically moved to an unnested format because it seemed cleaner / more convenient, and nesting just for evaluation might be as inefficient / RAM-exploding as a `group_split`.  (But perhaps if such nesting/`nest_by`ing is done before the join with the actuals, it might offer some offsetting performance gains.)  We talked a bit about this; do you remember if we/you made any conclusions about whether this was/wasn't worth pursuing?  (& if so, is it in an issue?)
* We still have the issue that: not every error measure needs all the `score_func_param_checker` checks, and these checks are repeated across the error measures. How best to avoid unnecessary and repeated checks isn't immediately obvious.  Is this in an issue somewhere?

**Comment by brookslogan** [comment]
From failing checks:
```
Err:33 https://ppa.launchpadcontent.net/cran/travis/ubuntu jammy Release
  404  Not Found [IP: 185.125.190.52 443]
Reading package lists...
E: The repository 'https://ppa.launchpadcontent.net/cran/travis/ubuntu jammy Release' does not have a Release file.
Repository: 'deb https://ppa.launchpadcontent.net/cran/travis/ubuntu/ jammy main'
Description:
Backports of important libraries used by R packages on Travis
More info: https://launchpad.net/~cran/+archive/ubuntu/travis
Adding repository.
Adding deb entry to /etc/apt/sources.list.d/cran-ubuntu-travis-jammy.list
Adding disabled deb-src entry to /etc/apt/sources.list.d/cran-ubuntu-travis-jammy.list
Adding key to /etc/apt/trusted.gpg.d/cran-ubuntu-travis.gpg with fingerprint 4A0C193118803EB4A561E569B3CF35C315B55A9F
/usr/bin/sudo DEBIAN_FRONTEND=noninteractive apt-get update -y -qq
E: The repository 'https://ppa.launchpadcontent.net/cran/travis/ubuntu jammy Release' does not have a Release file.
Error: Failed to get R 3.5.3: Failed to get R 3.5.3: Failed to install R: Error: The process '/usr/bin/sudo' failed with exit code 100
```

It looks like that [repo](https://ppa.launchpadcontent.net/cran/travis/ubuntu/dists/) does not have jammy releases yet.  Latest available seems to be [the previous LTS, focal](https://en.wikipedia.org/wiki/Ubuntu_version_history).  Or maybe `devel` could correspond to `jammy`??

File: R-packages/evalcast/R/error_measures.R [hunk 1]
```diff
@@ -201,41 +207,43 @@ sharpness <- function(quantile, value, actual_value) {
...
 is_symmetric <- function(x, tol=1e-8) {
-  x <- sort(x)
+  # Checking if `x` is sorted is much faster than trying to sort it again
+  if (is.unsorted(x, na.rm=TRUE)) {
+    # Implicitly drops NA values
+    x <- sort(x)
+  } else {
+    # Match `sort` behavior
+    x <- x[!is.na(x)]
+  }
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R-packages/evalcast/R/error_measures.R:242]: Yes, the amount of speedup is very surprising. It's somewhat less when the data doesn't include missing values, so `NA`-handling seems to be one of the slow parts.

**Comment by nmdefries** [comment]
You mentioned considering moving from assertthat to the checkmate package; what made you go with plain R instead?
What data/scripts were your benchmarks/profiling based on? Is the benchmarking code easy to share here in case we want to re-use it for future performance experiments?

**Comment by nmdefries** [comment]
> You mentioned considering moving from `assertthat` to the `checkmate` package; what made you go with plain R instead?

There were a few different reasons. Plain R was even faster. `checkmate` doesn't support custom error messages, so I'd have to catch and handle the default ones. And a lot of the `checkmate` functions are overkill compared to what we actually check. Seems that the `checkmate` functions are meant to check a whole host of characteristics of an obj (len, min/max, missings, etc) all at once.

> What data/scripts were your benchmarks/profiling based on? Is the benchmarking code easy to share here in case we want to re-use it for future performance experiments?

I have a series of not very organized R scripts. I'll clean up and see what might be useful for future testing.

> We still have the issue that: not every error measure needs all the `score_func_param_checker` checks, and these checks are repeated across the error measures. How best to avoid unnecessary and repeated checks isn't immediately obvious. Is this in an issue somewhere?

Will open an issue.

File: R-packages/evalcast/R/evaluate_predictions.R [hunk 2]
```diff
@@ -53,13 +56,17 @@ evaluate_predictions <- function(
...
-      lapply(erm, err_measures = err_measures) %>%
-      bind_rows() %>%
-      bind_cols(sc_keys) %>%
+    err_calls <- stats::setNames(
+      as.character(str_glue(
+        "err_measures[[{i}]](quantile, value, actual)",
+        i = 1:length(err_measures)
+      )),
+      names(err_measures))
+    err_calls <- parse_quos(err_calls, env=environment())
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R-packages/evalcast/R/evaluate_predictions.R:65]: Thank you for the suggestion, looks much much cleaner!

**Review by dshemetov — APPROVED** [review]
I don't have much to add here as I'm not too familiar with this code. Looks good from what I've seen.

**Pull request merged**